### DISCUSSION TITLE
Respect `Cache-Control` header for fetch data-cache

### DISCUSSION THREAD

**Discussion by rattrayalex** [vercel/next.js#53504] [Ideas]
### Goals

1. Allow library authors to disable caching in vercel's fetch data cache without leaking Next-specific types or runtime behavior.
2. Respect standards.

### Background

A user making requests with the openai TS SDK had to do some backflips to force Next to not use the cache for requests:

https://github.com/openai/openai-node/discussions/182#discussioncomment-6606706

We have workarounds available, but at least for certain endpoints, I'd expect ~nobody would want to cache these endpoints; it'd be nice if the library could either disable this cache by default, or allow the user to do so easily (passing headers is often easier through another library than nonstandard requestinit options).

### Proposal

I don't have specific proposals, but following https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching  more closely or at least respecting `Cache-Control: no-cache` would be great.

**Comment by rattrayalex** [comment]
cc @ijjk as https://github.com/vercel/next.js/pull/49638 looks relevant

**Comment by davidbarratt** [comment]
I recently read [the blog post](https://nextjs.org/blog/next-15-rc) for Next.js 15 RC and while the [new caching heuristics](https://nextjs.org/blog/next-15-rc#caching-updates) are a step in the right direction, It still leaves me puzzled how Next.js could come to the conclusion that the cache _instruction_ ought to be reversed.

The [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) header [was introduced in 1999](https://datatracker.ietf.org/doc/html/rfc2616) and has received [many updates and revisions](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#see_also) since its introduction.

What has remained constant since then is that the server has **always** instructed the client on how the response _can_ be cached.

Next.js reverses this basic assumption about HTTP cacheability and instead completely ignores how a server provides instructions on how a resource _can_ be cached by the client and instead imposes its own heuristics. This of course leads to severe under or over caching.

While the browser does [have a mechanism](https://developer.mozilla.org/en-US/docs/Web/API/Cache) for storing HTTP Responses manually, this is completely redundant if the server responds with a proper `Cache-Control` header. Of course, this mechanism could be useful for legacy or 3rd-party systems that are difficult or impossible to update. However, modern APIs _should_ respond with the proper response headers, and if they do not, they _ought_ to be updated to do so.

It seems sensible to me that Next.js could respect the `Cache-Control` header and act as a [shared cache](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#shared_cache) in the same way that [Varnish](https://varnish-cache.org/), [Squid](https://www.squid-cache.org/), [Cloudflare](https://developers.cloudflare.com/cache/), [Fastly](https://www.fastly.com/documentation/guides/concepts/edge-state/cache/), [Akamai](https://techdocs.akamai.com/api-definitions/docs/caching), and many others do.

Instead, Next.js simply ignores decades of existing software and charts its own path, one that makes little sense to this engineer who has been making websites since before the `Cache-Control` header was invented. I hope that Next.js changes this behavior and brings a consistent and coherent HTTP caching strategy to its users.
